Added LogFwk* logging levels to be used by the framework

Dr15Jones opened the pull request:
#### PR description:

- Added LogFwkInfo and LogFwkVerbatim with new level FWKINFO
- Refactored internals of MessageLogger to make it easier to maintain

This should make it easier to set a theshold such that INFO messages are ignored but the framework status messages are still kept.

#### PR validation:

Code compiles and framework unit tests pass.

Dr15Jones committed:
Made logging level object constexpr

Dr15Jones committed:
Consolidate all Log* classes into one implementation

Dr15Jones committed:
Added LogFwk* Options

The new options use the level fwkInfo which lies between info and warning.

Dr15Jones committed:
Use LogFwkVerbatim in the framework

-This required updating many test results

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
NOTE: I decided to wait for another pull request to actually use the new threshold as many non-framework configurations are dependent upon how the default MessageLogger config is structured right now.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19117

- This PR adds an extra 88KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629

cmsbuild commented:
A new Pull Request was created by  @Dr15Jones (Chris Jones) for master.

It involves the following packages:

FWCore/Framework
FWCore/MessageLogger
FWCore/MessageService
FWCore/Modules


@makortel, @smuzaffar, @cmsbuild, @Dr15Jones can you please review it and eventually sign? Thanks.
@makortel, @wddgit this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 09216a8b110550727909a239dabb91202afc3ca3

CMSSW: CMSSW_11_2_X_2020-10-15-1100
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-94af9a/10002/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  HeaderConsistency**

* **Build**:

I found compilation error when building: 

<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtHadEvtSolution.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtGenEvent.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtSemiLepEvtPartons.cc
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtSemiEvtSolution.cc
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtEvent.cc:3:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/FWCore/MessageLogger/interface/MessageLogger.h:124:42: error: conflicting declaration 'using LogInfo = class edm::Log<edm::level::Info, false>'
   using LogInfo = Log<level::Info, false>;
                                          ^
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/src/TtEvent.cc:1:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-10-15-1100/src/AnalysisDataFormats/TopObjects/interface/TtEvent.h:24:9: note: previous declaration as 'class edm::LogInfo'
   class LogInfo;
</pre>

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

Dr15Jones committed:
Made TtEvent::printParticle a stand alone function

Needed to remove forward declaration of edm::LogInfo since that
is no longer a class. The function did not depend on anything
in the class anyway.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19127

- This PR adds an extra 100KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629

cmsbuild commented:
Pull request #31815 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones, @santocch can you please check and sign again.

Dr15Jones commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on FWCore/MessageLogger/interface/MessageSender.h:
```diff
@@ -29,8 +18,10 @@ namespace edm {
+    MessageSender() = default;
     MessageSender(ELseverityLevel const& sev, std::string_view id, bool verbatim = false, bool suppressed = false);
+    MessageSender(MessageSender&&) = default;
+    MessageSender(MessageSender const&) = default;
```
How about defaulting or deleting the copy and move assignment operators explicitly?

Dr15Jones replied:
done

makortel commented on FWCore/MessageLogger/src/ELseverityLevel.cc:
```diff
@@ -150,6 +110,7 @@ namespace edm {
       ret[ELsev_zeroSeverity] = "--";
       ret[ELsev_success] = "Debug";  // 4/21/09 mf
```
Do you want to remove the comment? (I'm fine with either way)

Dr15Jones replied:
done

makortel commented on FWCore/MessageLogger/src/ELseverityLevel.cc:
```diff
@@ -188,6 +150,7 @@ namespace edm {
       ret[ELsev_zeroSeverity] = "ELzeroSeverity   ";
       ret[ELsev_success] = "ELdebug          ";  // 4/21/09
```
Do you want to remove the comment? (I'm fine with either way)

Dr15Jones replied:
done

Dr15Jones committed:
Explicitly tell compiler how to handle auto-generated functions

Dr15Jones committed:
Removed unnecessary comments

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19140

- This PR adds an extra 156KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629

cmsbuild commented:
Pull request #31815 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones, @santocch can you please check and sign again.

Dr15Jones commented:
NOTE: another small change is coming

Dr15Jones committed:
Make validation accept FWKINFO threshold

Dr15Jones commented:
The original pull request tests have succeeded in compiling the code and are now running the unit tests. I propose letting that test at least get through the unit test phase before launching a new set of tests on the updated code.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19143

- This PR adds an extra 156KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629

cmsbuild commented:
Pull request #31815 was updated. @makortel, @smuzaffar, @cmsbuild, @Dr15Jones, @santocch can you please check and sign again.

Dr15Jones commented:
Looking at the test log seems there were some failed tests
```
+ TEST_ERRORS='---> test unitTestsGroup_1 had ERRORS
---> test GeometryMTDGeometryBuilderTestDriver had ERRORS
---> test GeometryMTDNumberingBuilderTestDriver had ERRORS
---> test GeometryMTDCommonDataTestDriver had ERRORS'
```

cmsbuild commented:
-1

Tested at: 9924ab13d0565705b99f45dc3912698b82820833

CMSSW: CMSSW_11_2_X_2020-10-15-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-94af9a/10021/summary.html

I found follow errors while testing this PR

**Failed tests:  UnitTests**

* **Unit Tests**:

I found errors in the following unit tests: 
 
 ---> test unitTestsGroup_1 had ERRORS
---> test GeometryMTDGeometryBuilderTestDriver had ERRORS
---> test GeometryMTDNumberingBuilderTestDriver had ERRORS
---> test GeometryMTDCommonDataTestDriver had ERRORS

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-94af9a/10021/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 3 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543723
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

Dr15Jones committed:
Properly propagate FWKINFO within the MessageLogger config system

Dr15Jones committed:
Set limit on FWKINFO severity messages

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19153

- This PR adds an extra 180KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629


Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19153/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19153/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

Dr15Jones committed:
code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

Dr15Jones commented:
please test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-31815/19154

- This PR adds an extra 180KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  FWCore/Framework/src/InputSource.cc  modified in PR(s): #31658
  - File  FWCore/Framework/src/Schedule.cc  modified in PR(s): #29553
  - File  FWCore/MessageLogger/interface/MessageLogger.h  modified in PR(s): #30629
  - File  FWCore/MessageLogger/interface/MessageSender.h  modified in PR(s): #30629
  - File  FWCore/MessageService/src/MessageLogger.cc  modified in PR(s): #29553
  - File  FWCore/MessageService/test/unit_test_outputs/infos.log  modified in PR(s): #30629

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
Pull request #31815 was updated. @smuzaffar, @civanch, @Dr15Jones, @makortel, @cvuosalo, @ianna, @mdhildreth, @cmsbuild, @kpedro88, @santocch can you please check and sign again.

cmsbuild commented:
+1
Tested at: b2bd0aedbd6ad99e3fdaf0255e53658a9edd7f07
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-94af9a/10034/summary.html
CMSSW: CMSSW_11_2_X_2020-10-16-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-94af9a/10034/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2543752
 - DQMHistoTests: Total failures: 4
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2543726
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 149 log files, 22 edm output root files, 35 DQM output files

santocch commented:
+1

makortel commented:
+1

Dr15Jones commented:
@cms-sw/upgrade-l2 it would be great if you could sign this

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo, @qliphy (and backports should be raised in the release meeting by the corresponding L2)

qliphy commented:
+1
